src/foam/nanos/auth/ruler/AddPasswordHistoryAction.js

```diff
@@ -0,0 +1,61 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.auth.ruler',
+  name: 'AddPasswordHistoryAction',
+  documentation: 'Adds an entry into the password history of a user',
+
+  implements: ['foam.nanos.ruler.RuleAction'],
+
+  javaImports: [
+    'foam.nanos.logger.Logger',
+    'foam.nanos.auth.HistoricPassword',
+    'foam.nanos.auth.User',
+    'java.util.Date'
+  ],
+
+  constants: [
+    {
+      name: 'MAXIMUM_HISTORIC_PASSWORDS',
+      type: 'Long',
+      value: 10
+    }
+  ],
+
+  methods: [
+    {
+      name: 'applyAction',
+      javaCode: `
+        User user = (User) obj;
+        HistoricPassword[] historicPasswordsOld = user.getPasswordHistory();
+        
+        // compute how large the new array should be and where it should start copying from
+        int newArrayLength = historicPasswordsOld.length + 1;
+        int startCopyingIndex = 0;
+        int copyLength = historicPasswordsOld.length;
+        
+        // only store up to the maximum number of password history records
+        if (historicPasswordsOld.length >= MAXIMUM_HISTORIC_PASSWORDS) {
```

Comment:
Shouldn't this be

```suggestion
        if ( newArrayLength >= MAXIMUM_HISTORIC_PASSWORDS ) {
```
?

Reply:
yeah .. i thought of this too, but i need to be checking the old array length, not the new one. The idea here is that if the old array length is 9 (hence newArrayLength == 10), then i want to copy all 9 values (index 0-8), not 8 of them (index 1-8) .. so we need to check the old length.